src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs

```diff
@@ -92,14 +94,18 @@ private void Run()
             {
                 _delayStart.TrySetException(ex);
             }
+            _serviceStopped.TrySetResult(null);
         }
 
         public Task StopAsync(CancellationToken cancellationToken)
         {
-            // Avoid deadlock where host waits for StopAsync before firing ApplicationStopped,
-            // and Stop waits for ApplicationStopped.
-            Task.Run(Stop, CancellationToken.None);
-            return Task.CompletedTask;
+            if (!_isStopped)
+            {
+                Task.Run(Stop, CancellationToken.None);
```

Comment:
What do you want to have happen if Stop throws an exception?  Right now that exception is going to be eaten by the Task that's returned from Run and ignored (and at some point later the TaskSchedule.UnobservedTaskException event might be raised).

Reply:
It was previously ignored, so I left it that way.  I suppose it wouldn't hurt to throw here?  
This is the body of that task.  It would throw if for some reason it failed to set the service state to stopped.
https://github.com/dotnet/runtime/blob/3f5cb1805ea040e0869c1da8b8a7476a223ce4d2/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L502-L527
This is the caller:
https://github.com/dotnet/runtime/blob/3f5cb1805ea040e0869c1da8b8a7476a223ce4d2/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L149

To capture the exception would you reccomend `return Task.Run(Stop, CancellationToken.None).ContinueWith(_serviceStopped.Task)`?

Reply:
It depends on what we want to have happen.

If Stop throws an exception, does that mean _serviceStopped will never complete?  If that were the case someone doing `await StopAsync()` would hang, because the Task they're awaiting would never complete.  If it's really the case that either both complete successfully or at most one fails, then I think we want to do this:
```C#
public async Task StopAsync(CancellationToken cancellationToken)
{
    if (!isStopped)
    {
        await Task.Run(Stop, CancellationToken.None).ConfigureAwait(false);
    }

    return await _serviceStopped.Task;
}
```

Reply:
Got it.  This is where I landed as well after reding your comments.  Thanks for reminding me about `async` keyword to save my hackery with ContinueWith. 

I think I can actually force this path to throw as well, with user-code that overrides OnStop.  I'll write some tests for that.